File: backend/crdt/text.go
package crdt

import (
	"sort"
	"strings"
)

// TextCRDT is an RGA-flavoured sequence CRDT used for Docs (OFFICE-22).
//
// Each visible character is a node addressed by an OpID. An insert op
// places a new node "after" an existing parent OpID; concurrent inserts
// with the same parent are ordered deterministically by their OpID
// (Counter desc, then Replica desc) so every replica converges to the
// identical visible string.
//
// Deletes are tombstones: the node stays in the structure (so other
// inserts that referenced it still anchor correctly) but is hidden from
// the rendered text. This guarantees the structure is monotone, hence
// merge is commutative + idempotent.
type TextCRDT struct {
	nodes  map[OpID]*textNode
	root   OpID // sentinel ("zero" OpID) is the start anchor
	// childIndex maps parent OpID -> sorted child OpIDs (descending OpID).
	childIndex map[OpID][]OpID
}

type textNode struct {
	id      OpID
	parent  OpID // OpID this node was inserted *after*
	value   rune
	deleted bool
}

// NewTextCRDT returns an empty text document.
func NewTextCRDT() *TextCRDT {
	return &TextCRDT{
		nodes:      map[OpID]*textNode{},
		root:       OpID{},
		childIndex: map[OpID][]OpID{},
	}
}

// TextOpKind is the union of mutating ops on a text CRDT.
type TextOpKind uint8

const (
	TextOpInsert TextOpKind = 1
	TextOpDelete TextOpKind = 2
)

// TextOp is a single text mutation. Serialized into Op.Body when sent
// over a Transport.
type TextOp struct {
	Kind   TextOpKind `json:"k"`
	ID     OpID       `json:"id"`
	Parent OpID       `json:"p,omitempty"` // for Insert: anchor; for Delete: target id is in Target
	Value  rune       `json:"v,omitempty"`
	Target OpID       `json:"t,omitempty"` // for Delete
}

// Apply applies op idempotently. It is safe to apply the same op twice
// or in any order with respect to other ops. The only requirement is
// that an insert's Parent has been observed before the insert is
// applied (causal delivery); the Document layer enforces this by
// buffering ops with missing causes.
func (t *TextCRDT) Apply(op TextOp) {
	switch op.Kind {
	case TextOpInsert:
		if _, ok := t.nodes[op.ID]; ok {
			return // idempotent
		}
		t.nodes[op.ID] = &textNode{
			id:     op.ID,
			parent: op.Parent,
			value:  op.Value,
		}
		t.insertChild(op.Parent, op.ID)
	case TextOpDelete:
		// Doc layer (Doc.textCauseReady) buffers deletes whose target
		// isn't observed yet, so by the time we get here Target should
		// exist. Tolerate the rare case where it doesn't.
		n, ok := t.nodes[op.Target]
		if !ok {
			return
		}
		n.deleted = true
	}
}

// insertChild inserts cid into parent's child list, kept sorted in
// descending OpID order (so a left-to-right depth-first walk produces
// the correct RGA visible order).
func (t *TextCRDT) insertChild(parent, cid OpID) {
	children := t.childIndex[parent]
	idx := sort.Search(len(children), func(i int) bool {
		// descending: we want the position where children[i] < cid
		return children[i].Less(cid)
	})
	children = append(children, OpID{})
	copy(children[idx+1:], children[idx:])
	children[idx] = cid
	t.childIndex[parent] = children
}

// String renders the visible string. Stable across replicas.
func (t *TextCRDT) String() string {
	var b strings.Builder
	t.walk(t.root, &b)
	return b.String()
}

func (t *TextCRDT) walk(parent OpID, b *strings.Builder) {
	for _, cid := range t.childIndex[parent] {
		n := t.nodes[cid]
		if n != nil && !n.deleted && n.value != 0 {
			b.WriteRune(n.value)
		}
		t.walk(cid, b)
	}
}

// VisibleIDs returns the visible character OpIDs in document order.
// Useful for resolving a UI cursor index to a stable parent anchor.
func (t *TextCRDT) VisibleIDs() []OpID {
	var out []OpID
	t.collect(t.root, &out)
	return out
}

func (t *TextCRDT) collect(parent OpID, out *[]OpID) {
	for _, cid := range t.childIndex[parent] {
		n := t.nodes[cid]
		if n != nil && !n.deleted && n.value != 0 {
			*out = append(*out, cid)
		}
		t.collect(cid, out)
	}
}

// LocalInsert is a helper that returns the TextOp for inserting `r`
// after visible index `i` (0 = beginning). Caller is responsible for
// stamping the OpID via the document's Lamport clock.
func (t *TextCRDT) LocalInsert(i int, r rune, id OpID) TextOp {
	parent := t.root
	if i > 0 {
		vis := t.VisibleIDs()
		if i-1 < len(vis) {
			parent = vis[i-1]
		} else if len(vis) > 0 {
			parent = vis[len(vis)-1]
		}
	}
	return TextOp{Kind: TextOpInsert, ID: id, Parent: parent, Value: r}
}

// LocalDelete returns the TextOp for deleting the character at visible
// index `i`.
func (t *TextCRDT) LocalDelete(i int, id OpID) (TextOp, bool) {
	vis := t.VisibleIDs()
	if i < 0 || i >= len(vis) {
		return TextOp{}, false
	}
	return TextOp{Kind: TextOpDelete, ID: id, Target: vis[i]}, true
}

// snapshot exports the CRDT state for bucket persistence.
type textSnapshot struct {
	Nodes []textNodeSnapshot `json:"nodes"`
}

type textNodeSnapshot struct {
	ID      OpID `json:"id"`
	Parent  OpID `json:"p"`
	Value   rune `json:"v"`
	Deleted bool `json:"d,omitempty"`
}

func (t *TextCRDT) snapshot() textSnapshot {
	out := textSnapshot{Nodes: make([]textNodeSnapshot, 0, len(t.nodes))}
	for _, n := range t.nodes {
		out.Nodes = append(out.Nodes, textNodeSnapshot{
			ID: n.id, Parent: n.parent, Value: n.value, Deleted: n.deleted,
		})
	}
	// Deterministic ordering for stable hashing of snapshots.
	sort.Slice(out.Nodes, func(i, j int) bool {
		return out.Nodes[i].ID.Less(out.Nodes[j].ID)
	})
	return out
}

func (t *TextCRDT) restore(s textSnapshot) {
	t.nodes = map[OpID]*textNode{}
	t.childIndex = map[OpID][]OpID{}
	// Apply parent-first so child lists are built consistently. We
	// approximate by sorting by OpID ascending; since Counter only
	// grows, parents arrive before children.
	sort.Slice(s.Nodes, func(i, j int) bool { return s.Nodes[i].ID.Less(s.Nodes[j].ID) })
	for _, n := range s.Nodes {
		t.nodes[n.ID] = &textNode{
			id: n.ID, parent: n.Parent, value: n.Value, deleted: n.Deleted,
		}
		t.insertChild(n.Parent, n.ID)
	}
}

File: src/lib/crdt/text.js
/**
 * src/lib/crdt/text.js
 *
 * Browser-side RGA text CRDT for Vulos Office (OFFICE-22).
 *
 * Mirrors backend/crdt/text.go and backend/crdt/id.go so that ops
 * encoded here are accepted by the Go Doc.ApplyRemote() and vice-versa.
 *
 * Wire format (JSON) matches the Go TextOp / OpID structs:
 *   { k: 1|2, id: {r,c}, p: {r,c}, v: <codepoint>, t: {r,c} }
 *   k=1 → Insert  k=2 → Delete
 *
 * Usage:
 *   const crdt = new TextCRDT('replica-a')
 *   const op = crdt.localInsert(0, 'H')   // insert 'H' at position 0
 *   crdt.apply(op)                         // apply locally
 *   // broadcast op to peers …
 *   crdt.apply(remoteOp)                   // merge remote op
 *   crdt.toString()                        // → current visible string
 */

// ---------------------------------------------------------------------------
// Lamport clock (matches Go LamportClock)
// ---------------------------------------------------------------------------

class LamportClock {
  constructor(replica) {
    this.replica = replica
    this.c = 0
  }

  tick() {
    this.c += 1
    return { r: this.replica, c: this.c }
  }

  observe(remoteC) {
    if (remoteC > this.c) this.c = remoteC
  }
}

// ---------------------------------------------------------------------------
// OpID helpers (match Go OpID.Less / Equal)
// ---------------------------------------------------------------------------

function opidLess(a, b) {
  if (a.c !== b.c) return a.c < b.c
  return a.r < b.r
}

function opidEqual(a, b) {
  return a.c === b.c && a.r === b.r
}

function opidZero(id) {
  return !id || (id.c === 0 && (!id.r || id.r === ''))
}

const ROOT_ID = { r: '', c: 0 }

// ---------------------------------------------------------------------------
// Op kind constants (match Go TextOpInsert / TextOpDelete)
// ---------------------------------------------------------------------------

export const TEXT_OP_INSERT = 1
export const TEXT_OP_DELETE = 2

// ---------------------------------------------------------------------------
// TextCRDT
// ---------------------------------------------------------------------------

export class TextCRDT {
  /**
   * @param {string} replicaId  - unique stable id for this peer
   */
  constructor(replicaId) {
    this._replica = replicaId
    this._clock = new LamportClock(replicaId)

    // Map<opid-key, node>  where node = { id, parent, value, deleted }
    this._nodes = new Map()
    // Map<opid-key, opid[]>  child lists kept sorted descending (RGA order)
    this._children = new Map()

    this._rootKey = _key(ROOT_ID)
    this._children.set(this._rootKey, [])
  }

  // ─── Local mutation helpers ─────────────────────────────────────────────

  /**
   * Produce an Insert op for character `ch` at visible index `i`
   * (0 = before all existing text).  Does NOT apply the op; caller
   * must call apply() after broadcasting.
   */
  localInsert(i, ch) {
    const id = this._clock.tick()
    const parent = this._parentAt(i)
    return { k: TEXT_OP_INSERT, id, p: parent, v: ch.codePointAt(0) }
  }

  /**
   * Produce a Delete op for the character at visible index `i`.
   * Returns null if the index is out of range.
   */
  localDelete(i) {
    const vis = this._visibleIds()
    if (i < 0 || i >= vis.length) return null
    const id = this._clock.tick()
    return { k: TEXT_OP_DELETE, id, t: vis[i] }
  }

  // ─── Apply (local + remote, idempotent) ─────────────────────────────────

  /**
   * Apply a TextOp produced locally or received from a peer.
   * Safe to call multiple times for the same op (idempotent).
   * Returns true if the op changed the document state.
   */
  apply(op) {
    if (op.k === TEXT_OP_INSERT) {
      const k = _key(op.id)
      if (this._nodes.has(k)) return false  // already seen

      this._clock.observe(op.id.c)
      const parent = op.p && !opidZero(op.p) ? op.p : ROOT_ID
      const parentKey = _key(parent)

      this._nodes.set(k, { id: op.id, parent, value: String.fromCodePoint(op.v), deleted: false })
      if (!this._children.has(parentKey)) this._children.set(parentKey, [])
      this._insertChild(parentKey, op.id)
      return true
    }

    if (op.k === TEXT_OP_DELETE) {
      const targetKey = _key(op.t)
      const node = this._nodes.get(targetKey)
      if (!node || node.deleted) return false
      this._clock.observe(op.id.c)
      node.deleted = true
      return true
    }

    return false
  }

  // ─── Read ───────────────────────────────────────────────────────────────

  /** Returns the visible text string. */
  toString() {
    const parts = []
    this._walk(this._rootKey, parts)
    return parts.join('')
  }

  /**
   * Returns an array of opids for currently-visible characters in order.
   * Used to resolve a cursor offset to a stable anchor before broadcasting.
   */
  visibleIds() {
    return this._visibleIds()
  }

  // ─── Snapshot / restore (for bucket persistence) ────────────────────────

  /**
   * Export a plain-object snapshot of the full CRDT state.
   * Equivalent to Go TextCRDT.snapshot().
   */
  snapshot() {
    const nodes = []
    for (const [, node] of this._nodes) {
      nodes.push({ id: node.id, p: node.parent, v: node.value.codePointAt(0), d: node.deleted })
    }
    // Stable deterministic order by OpID ascending (matches Go).
    nodes.sort((a, b) => opidLess(a.id, b.id) ? -1 : 1)
    return { nodes }
  }

  /**
   * Restore state from a snapshot (replaces current state).
   */
  restore(snap) {
    this._nodes.clear()
    this._children.clear()
    this._children.set(this._rootKey, [])

    if (!snap || !Array.isArray(snap.nodes)) return

    // Sort ascending so parents are inserted before children.
    const sorted = [...snap.nodes].sort((a, b) => opidLess(a.id, b.id) ? -1 : 1)
    for (const n of sorted) {
      const k = _key(n.id)
      const parent = n.p && !opidZero(n.p) ? n.p : ROOT_ID
      const parentKey = _key(parent)
      this._nodes.set(k, { id: n.id, parent, value: String.fromCodePoint(n.v), deleted: !!n.d })
      if (!this._children.has(parentKey)) this._children.set(parentKey, [])
      this._insertChild(parentKey, n.id)
    }
  }

  // ─── Internal helpers ───────────────────────────────────────────────────

  _parentAt(i) {
    if (i <= 0) return ROOT_ID
    const vis = this._visibleIds()
    if (vis.length === 0) return ROOT_ID
    if (i - 1 < vis.length) return vis[i - 1]
    return vis[vis.length - 1]
  }

  _visibleIds() {
    const out = []
    this._collectVisible(this._rootKey, out)
    return out
  }

  _collectVisible(parentKey, out) {
    const children = this._children.get(parentKey) || []
    for (const cid of children) {
      const k = _key(cid)
      const node = this._nodes.get(k)
      if (node && !node.deleted) out.push(cid)
      this._collectVisible(k, out)
    }
  }

  _walk(parentKey, parts) {
    const children = this._children.get(parentKey) || []
    for (const cid of children) {
      const k = _key(cid)
      const node = this._nodes.get(k)
      if (node && !node.deleted) parts.push(node.value)
      this._walk(k, parts)
    }
  }

  /**
   * Insert cid into parent's child list keeping it sorted in descending
   * OpID order (mirrors Go TextCRDT.insertChild).
   */
  _insertChild(parentKey, cid) {
    const children = this._children.get(parentKey)
    // Find insertion index: first position where children[i] < cid (descending).
    let idx = 0
    while (idx < children.length && !opidLess(children[idx], cid)) idx++
    children.splice(idx, 0, cid)
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function _key(id) {
  if (!id || opidZero(id)) return '__root__'
  return `${id.r}@${id.c}`
}
